### ISSUE TITLE
Locking / resolving a version without appropriate distribution wheels available

### ISSUE THREAD

**Issue by jonaslb** [astral-sh/uv#9412]
Currently, the `eccodes` library's latest release, v. 2.39.0, is released without source or manylinux builds. It's an error on their side for sure, and I've created an issue in their repo. However, in spite of this, `uv` still resolves this latest problematic version when locking, and it means that `eccodes` can't be added without version specifier:

```sh
$ uv add eccodes
Resolved 7 packages in 821ms
error: Distribution `eccodes==2.39.0 @ registry+https://pypi.org/simple` can't be installed because it doesn't have a source distribution or wheel for the current platform
```

Specifying the old version (which has the right files on PyPI) works fine. However, in my projects I don't depend directly on eccodes, it's somewhere down the dependency tree, so I'd prefer not having to explicitly pin it in my pyproject file.

So, I'm not sure if it should be considered a bug in uv, but it's certainly annoying that a dependency somewhere in the tree can create this kind of disturbance.

Likely related (don't know if considered dupe):
- https://github.com/astral-sh/uv/issues/7816
- https://github.com/astral-sh/uv/issues/8375
- https://github.com/astral-sh/uv/issues/8492

**Comment by charliermarsh** [comment]
Ah yeah, this is expected, though it's understandably not a great experience. It's another case of the problem I outlined here: https://github.com/astral-sh/uv/issues/5182#issuecomment-2237858277. I'd prefer to track in that issue. I think right now, you may want to set a constraint on your end (with `constraint-dependencies`).

**Comment by jonaslb** [comment]
Returning to this because I just had the issue again with another ecmwf library, `eckitlib`, which in the latest version only has macos wheels. Since #5182 was closed, I believe the new tracking issue is #9711.

**Comment by konstin** [comment]
uv now supports [required environments](https://docs.astral.sh/uv/concepts/resolution/#required-environments) for which wheels must exist for a version to be selected, this should match the original problem well.
